### ISSUE TITLE
Setting preexec_fn to call os.setpgrp() breaks scripts calling matlab

### ISSUE THREAD

**Issue by bjeurissen** [MRtrix3/mrtrix3#1457]
On linux, setting `preexec_fn` to call `os.setpgrp()` in `run.py` caused all my scripts that call matlab to break.

Calling matlab directly using `subprocess` (without `preexec_fn`) would fix the scripts themselves, but when calling those scripts from yet another script using `run.command` would fail again.

Removing the `preexec_fn` option ensured that all scripts directly or indirectly calling matlab, work fine.

Calling matlab with `-singleCompThread` and `-nojvm` also did not help.

**Comment by Lestropie** [comment]
OK, this'll be a good bit of fun to find the right configuration :-/  I had to muck around quite a bit to try to have system signals handled appropriately in all possible scenarios, but I never fully documented the precise logic that led to the final solution. Having said that I've never been wholly convinced that I got it right...

Traceback:
467699fc dealt with `os.setpgrp()` not being defined on Windows.
2956b244 added the `os.setpgrp()` call. 

So basically what we need is:
- Ctrl-C should call the custom signal handler and appropriately terminate any & all child threads.
- This should happen regardless of whether a child thread is currently executing as part of a `run.command()` call, or whether script execution is currently within Python code.
- Should work on both Unix and Windows.
- Should work on both Python 2 and 3.
- Should not use `os.setpgrp()` on Unix.
:frowning_face: :gun:

**Comment by jdtournier** [comment]
I'm not sure I get all the subtleties here... But just from a quick look into this at the time:

-  Wouldn't [CREATE_NEW_PROCESS_GROUP](https://docs.python.org/2/library/subprocess.html#subprocess.CREATE_NEW_PROCESS_GROUP) be more appropriate on Windows?

- Is there a problem on Linux if we _don't_ call `os.setpgrp()`? It certainly seemed to cause trouble here, but you state it shouldn't be called, yet it most certainly is being called - and yet it's not called on Windows either... I'm confused.

- Maybe the answer might be to invoke the default signal handler explicitly (or re-raise the exception, if such a thing is possible?) after the required cleanup, to make sure that all this is handled as expected - at least on Unix?

**Comment by Lestropie** [comment]
> I'm not sure I get all the subtleties here...

I probably don't get them all either; hence why the whole thing needs to be approached fresh. I seem to recall that the signal handling in the Python libraries actually wasn't quite working properly at first, which may have led to improper solutions in the surrounding code.

> Wouldn't CREATE_NEW_PROCESS_GROUP be more appropriate on Windows?

This would mean that system signals generated would not automatically propagate to all processes. Now that's not necessarily a bad thing, but all of the surrounding code would need to be designed with this in mind. Note also that POSIX compliance for particular behaviours may or may not occur on Windows, hence why solutions need to be tested on all platforms.

> Is there a problem on Linux if we don't call `os.setpgrp()`? It certainly seemed to cause trouble here, but you state it shouldn't be called, yet it most certainly is being called - and yet it's not called on Windows either... I'm confused.

It certainly caused a problem on *something* if it wasn't there, hence why it's there. When I state "it shouldn't be called", I don't mean that the current code is not supposed to call it: I mean that any new solution should not invoke that function on Linux since it's been observed to cause an issue.

> Maybe the answer might be to invoke the default signal handler explicitly (or re-raise the exception, if such a thing is possible?) after the required cleanup, to make sure that all this is handled as expected - at least on Unix?

Maybe. There may also be differential handling required between Ctrl-C's and other signals, particularly since the former is dealt with differently on Windows, hence the explicit [catching](https://github.com/MRtrix3/mrtrix3/blob/master/lib/mrtrix3/run.py#L174-L175) in the current code.

Sorry if I sound slightly averse to suggestions: in the past this has taken a lot of gymnastics to get anywhere resembling desired behaviour, so I'm by default sceptical of anything that hasn't been tested against every possible use case.

**Comment by jdtournier** [comment]
OK, thanks for clarifying. I guess it would be helpful to set up some kind of minimal working example so we can check all this out. I'm a bit unsure as to what the relevant scenarios are, but I'm going to list what I think we're after below, please correct if wrong:

If we receive a signal during execution of one of our scripts: 
1. any child process should also be terminated appropriately
2. we need to make sure any descendants of these processes are also appropriately terminated
3. main script should invoke clean-up function to delete temporaries if a signal or exception is caught.

My guess is we have to do things in this order to ensure that the temporaries we are deleting aren't concurrently being written to by child processes. On Unix, this will probably be handled appropriately (the child process will write to an unlinked file, which will be deleted properly once all file handles to it have been closed). Unlikely to work on Windows, it invariably prevent opening, modifying or deleting files that are currently open in other processes. So for this to work, I assume we need to make sure everything happens in that order. Which kind of kills the idea of invoking clean-up, then invoking default system handler... 

Also potentially related: on Windows, Ctrl-C on the terminal often hangs waiting for some of the child processes to complete - there is no attempt at terminating them, as far as I can tell. This is most obvious during `./build`, which does not use custom signal handling. However, this might be related to multi-threading rather than separate processes, and seems to be a [known issue on Windows](https://bugs.python.org/issue21822), with no fix yet as far as I can tell. Not sure how relevant it is to this discussion...

**Comment by jdtournier** [comment]
I've just had a go at putting together a minimal script to figure this out, not sure I'm on the right track though. I've pushed this to the `investigate_signal_handling` branch (based on `dev`). It add a `bin/test_signal_handling` script, which is a bare-bones app, disables the setting of the signals in the main `app.init()` call, and sets up a locally-defined handler for `SIGINT`. It then calls 3 instances of a bash script that simply echoes 'start PID', sleeps for 3 seconds, then echoes 'stop PID' so we can check if it's completed. Unfortunately, it all seems to work OK, although for some reason on my Win10 laptop it makes no difference whether I manually terminate the child processes or not - they're marked `<defunct>` straight away (as reported by `ps aux`). I have a feeling I'm missing something important somewhere...
